### DISCUSSION TITLE
how to Flow in PhysX

### DISCUSSION THREAD

**Discussion by oneanime** [NVIDIA-Omniverse/PhysX#325] [Q&A]
I am trying to simulate fluids using PhysX and accelerate the process on the GPU. However, I couldn't find any examples or demos in the flow folder.

**Comment by vreutskyy** [comment]
Hi @oneanime,

PhysX and Flow are separate libraries, although they share the same repository. Flow is not actually part of PhysX.

You can find Flow's documentation here: [NVIDIA Flow Documentation](https://nvidia-omniverse.github.io/PhysX/flow/index.html).

On the other hand, PhysX has its own method for simulating particle-based fluids. For examples, you can check out SnippetPBF and other related snippets in the PhysX SDK. You can find the documentation for PhysX's particle systems here: [PhysX Particle Systems Documentation](https://nvidia-omniverse.github.io/PhysX/physx/5.4.2/docs/ParticleSystem.html#particle-system).

**Reply by oneanime** [reply]
And
Thank you for the amazing project! I'm currently studying physics-based simulations and I wanted to ask if the fluid simulation in this project is based on the SPH (Smoothed Particle Hydrodynamics) algorithm? If not, could you briefly explain which method is used for the fluid simulation?

**Reply by vreutskyy** [reply]
Hi @oneanime,

Thank you for your interest in the project! The fluid simulation in PhysX is not based on the traditional SPH (Smoothed Particle Hydrodynamics) algorithm. Instead, it uses a more advanced method called Position Based Fluids (PBF).

PBF was introduced in this paper: [Position Based Fluids](https://mmacklin.com/pbf_sig_preprint.pdf) by Macklin and Müller.

Here are some key points about the fluid simulation in PhysX:

1. It's based on Position Based Dynamics (PBD), with PBF being an extension specifically designed for fluid simulation.

2. PBF improves upon SPH in several ways:
   - It's more stable, allowing for larger time steps
   - It better preserves fluid volume
   - It handles surface tension more effectively

3. The main difference is that PBF solves for particle positions directly, rather than computing forces and then integrating to get positions. This approach makes it more stable and easier to constrain.

4. PhysX's implementation also includes further optimizations beyond the basic PBF algorithm to improve performance and visual quality.

If you're studying physics-based simulations, I'd recommend reading the paper I linked above. It provides a great introduction to the PBF method and its advantages over traditional SPH.

**Reply by oneanime** [reply]
We are working on integrating our custom fluid dynamics and collision resolution algorithms into PhysX. Could you please let us know if PhysX provides any interfaces or extension points for such integrations?

**Reply by vreutskyy** [reply]
Hi @oneanime 

PhysX does offer some interfaces and extension points for integrating custom algorithms, but there are limitations depending on the specific features you need.

**Fluid Dynamics**: PhysX primarily supports fluid simulation through GPU-accelerated features, which are not part of the open-source CPU code. These fluid models and advanced particle-based dynamics (PBD) are incorporated from the former NVIDIA Flex library and are available only through pre-compiled GPU binaries.

**Collision Resolution**: PhysX includes robust support for rigid body dynamics, including collision detection and resolution. The rigid body constraint solver is available in the CPU source code, but the GPU-accelerated version is part of the closed-source GPU code.

If your custom algorithms are designed to work with GPU-accelerated features, you might need to integrate them with the pre-compiled GPU binaries provided by NVIDIA. For CPU-based implementations, you can extend the existing PhysX source code available on GitHub.

**Reply by oneanime** [reply]
Thank you for providing such an excellent framework for physics simulation. We truly appreciate the effort and dedication that has gone into creating and maintaining PhysX. Before we proceed with posting on GitHub, we would like to clarify a few questions regarding PhysX and NVIDIA Flex. We would be grateful if you could provide some guidance on the following:

1. NVIDIA Flex Open Source Status:
From what we have observed on GitHub, it seems that the NVIDIA Flex codebase has not been actively maintained for quite some time. Could you confirm whether the entire NVIDIA Flex code is open source? If so, are we allowed to modify the Flex code and re-integrate it into PhysX for our custom use cases?

2. Handling Fluid-Solid Interactions in PhysX:
When PhysX handles scenarios involving both fluids and rigid bodies, how is the interaction between these two simulated? Specifically, does PhysX calculate the fluid dynamics first and then pass the data to the solid solver (or vice versa), or does it solve the coupled fluid and solid interactions simultaneously in a coupled manner? Understanding this would greatly help us in evaluating its suitability for our project.

3. Handing rigid body dynamics：
Can PhysX's rigid body contact detection or the constraint solver be replaced by our own GPU code for different algorithm from PhysX? Can we integrate our own algorithm code while keeping other PhysX modules in use?

We are looking forward to your response and would greatly appreciate any advice or guidance you can provide. Thank you again for your hard work and for making PhysX accessible to the community.

**Reply by vreutskyy** [reply]
Hi @oneanime 

I'll try to address your questions to the best of my understanding:

**NVIDIA Flex Open Source Status**
The entire NVIDIA Flex codebase is indeed open source and available on GitHub. You are allowed to modify the Flex code and re-integrate it into PhysX for your custom use cases, subject to the terms of its license. However, please note that Flex is a legacy SDK and is no longer actively supported by NVIDIA.

**Handling Fluid-Solid Interactions in PhysX**
PhysX handles fluid-solid interactions in a fully coupled manner. The simulation solves the interactions between fluids and rigid bodies simultaneously, allowing for two-way interaction between these elements. This approach ensures accurate and realistic simulations of complex scenarios involving both fluids and solids.

**Handling Rigid Body Dynamics**
Regarding the replacement of PhysX's rigid body contact detection or constraint solver with your own GPU code, it would be challenging to replace or expand PhysX's GPU features, as the GPU code is not open-sourced. Integrating your own algorithm code while keeping other PhysX modules in use may be difficult due to the closed-source nature of the GPU components.

For any specific implementation details or further clarifications, I recommend consulting the official PhysX documentation.

**Reply by lanhung** [reply]
@vreutskyy What's your suggestion if we want to implement our own Fluid Dynamics, e.g. LBM, and integrate it in Physx, instead of using the PBF, for the purpose of handling the fluid-solid interaction? Can we package our own algorithms and specify it being used in Physx? Thanks.
